src/interface/var_id.hpp

```diff
@@ -0,0 +1,56 @@
+//========================================================================================
+// (C) (or copyright) 2020-2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_VAR_ID_HPP_
+#define INTERFACE_VAR_ID_HPP_
+
+#include <functional>
+#include <string>
+
+namespace parthenon {
+
+inline std::string MakeVarLabel(const std::string &base_name, int sparse_id) {
+  return base_name +
+         (sparse_id == InvalidSparseID ? "" : "_" + std::to_string(sparse_id));
+}
+
+/// We uniquely identify a variable by it's full label, i.e. base name plus sparse ID.
+/// However, sometimes we also need to be able to separate the base name from the sparse
+/// ID. Instead of relying on the fact that they are separated by a "_", we store them
+/// separately in VarID struct. This way we know that a dense variable "foo_3" does not
+/// have a sparse ID and a sparse field "foo_3" has base name "foo" and sparse ID 3,
+/// however, the two VarIDs representing them are still considered equal, so that we find
+/// such duplicates
+/// TODO(JMM): Using VarID machinery for prolongation/restriction
+/// implies that all vars in a sparse pool have the same custom
+/// prolongation/restriction operators.
+struct VarID {
```

Comment:
I like moving this into its own file. :+1:

Reply:
Yeah, the hardest part about this PR was fixing circular dependencies.